### ISSUE TITLE
Automatic assets optimization

### ISSUE THREAD

**Issue by tua-Mascot** [foundryvtt/foundryvtt#10491]
### User Experience

Hello everyone.
Every web developer understands why it is important to optimize assets. However, Foundry VTT is not only used by developers but also by regular users who may not know anything about it. Even if you show them the optimization section on the Foundry website (https://foundryvtt.com/article/media/), most people still won't understand why it's necessary. In their worldview, it may seem like an unnecessary step. As a result, these people don't see anything wrong with adding a 10,000 x 10,000 px resolution PNG image or other non-optimized assets. Currently, I have to manually run my optimization script and pray to all gods that nothing breaks. 
I believe that the functionality of automatic assets optimization should be available out of the box, and no one but the Foundry developers can do it better (as you have a better understanding of how your project works).

I would like to see the following features:
1. Mass optimization of already added assets according to specified rules.
2. Optimization of assets at the time of their addition according to specified rules.

For example:
1. For images, you can set the maximum weight, maximum resolution, and preferred format (output). For example, if a user wants to upload a previously mentioned 10,000 x 10,000 px PNG file, Foundry optimizes it to 2,000 x 2,000 WebP.
2. For audio files, you can set the maximum weight and maximum kbps. For example, if a user wants to upload a 320 kbps MP3 file, Foundry optimizes it to 192 kbps MP3.
And so on. I hope I conveyed the general idea.

For assets optimization, I recommend using:
1. For classic image formats - sharp (https://github.com/lovell/sharp)
2. For SVG - svgo (https://github.com/svg/svgo)
3. For audio and video formats - ffmpeg (https://www.ffmpeg.org/)

PS
I'll allow myself a small off-topic. It doesn't directly relate to the above topic but concerns optimization in general. In the documentation (https://foundryvtt.com/article/configuration/ - The Configuration Menu), there is a section called "Compress Static Files." The description says, "When checked, Foundry will attempt to minify and compact all files (over 10 KB) sent to clients when they connect." Perhaps I misunderstood the word "minify" in this context, but even with this option enabled, I still see some JS files being transmitted with a large number of JSDoc comments (for example, /scripts/foundry.js). This doesn't seem like modern minification. Maybe this option is not working correctly, or the documentation needs a more precise description.

I apologize if this proposal already exists and has been discussed, but a quick search on issues did not yield any results.
In conclusion, I want to say thank you to all the Foundry developers and the community for your work 🤝.

**Comment by aaclayton** [comment]
I appreciate your thoughtfully worded suggestion. I want to be transparent that I doubt we will do this, I view it as the role of the Foundry VTT user to manage their own input assets - in fact I view that as one of the greatest strengths of the platform that it *does not* try to make or force these sorts of decisions for you.

My mind has changed on similar issues in the past though, keeping this open for long-term consideration.

**Comment by MrVauxs** [comment]
With the popularity of modules that auto-compress uploaded files and general heightened awareness of the average foundry user about performance bottlenecks (animations, large maps, cleaning your chat logs, putting stuff into compendiums), I genuinely think Foundry would benefit from a built-in setting to automatically compress everything to its optimal version. Even just lossless `.webp` compression does wonders. 

The action could be as simple as a toggle next to the upload button that persists with user settings. The main problem is not users not wanting it, but it being too bothersome. Barely anyone will be mad about being able to 10x their effective hard drive space at the cost of ticking a checkbox and Foundry having files in a different format than what you put in.

**Comment by aaclayton** [comment]
A friction point is doing cross-platform/architecture image manipulation on the Electron server-side. There are libraries but many of them have very complicated dependency chains or OS dependencies which make cross-platform builds problematic.

**Comment by MrVauxs** [comment]
Is there a necessity for doing it server-side? Client-side compression already exists through workers and we are talking about users who already have file permissions. And more often than not the server is about as or less powerful than the user machine.
